### ISSUE TITLE
ACL replication issue

### ISSUE THREAD

**Issue by ccakes** [hashicorp/consul#6431]
#### Overview of the Issue

I have a cluster up and running happily in AWS eu-west-1 and it's been there for around a year, no issues. I'm a little behind up patches atm but it's been upgraded and restarted multiple times with no issue

Today I created a new cluster in us-east-1 and tried joining them following the guides: [multi-dc](https://learn.hashicorp.com/consul/security-networking/datacenters) and [acl replication](https://learn.hashicorp.com/consul/day-2-operations/acl-replication)

I've been doing some test runs in a stage environment and it seems that ACLs are replicated so rather than create any on the new cluster, I just bootstrapped a master, set the replication token that I'd created on the existing cluster and then planned on using the ACLs that'd be replicated (when that happened).

They don't seem to be being replicated properly. API output from both clusters is:
```
# Existing (primary DC)
$ curl http://localhost:8500/v1/acl/replication?pretty
{
    "Enabled": false,
    "Running": false,
    "SourceDatacenter": "",
    "ReplicationType": "",
    "ReplicatedIndex": 0,
    "ReplicatedRoleIndex": 0,
    "ReplicatedTokenIndex": 0,
    "LastSuccess": "0001-01-01T00:00:00Z",
    "LastError": "0001-01-01T00:00:00Z"
}

# New (secondary DC)
$ curl http://localhost:8500/v1/acl/replication?pretty
{
    "Enabled": true,
    "Running": true,
    "SourceDatacenter": "aws-dln",
    "ReplicationType": "tokens",
    "ReplicatedIndex": 0,
    "ReplicatedRoleIndex": 0,
    "ReplicatedTokenIndex": 0,
    "LastSuccess": "0001-01-01T00:00:00Z",
    "LastError": "0001-01-01T00:00:00Z"
}
```
The output from the existing cluster looks bad, but both have `acl { enable_token_replication = true }` in the config so I'm guessing that just means it's not actively replicating from a remote source?

I note that the ACL replication guide mentions this:
> Note, your initial bootstrap token can be used for the necessary privileges to complete any action on the secondary servers.

If I run `consul acl token list` on the **new** cluster with it's own token then I get an error
```
export CONSUL_HTTP_TOKEN=<snip>
[root@ip-172-16-3-192 ~]# consul acl token list
Failed to retrieve the token list: Unexpected response code: 403 (ACL not found)
```
but if I do the same with a valid token **from the old** cluster, I see the list of tokens there. I assume this is the local Consul server forwarding the request but it's happening transparently so not sure

#### Reproduction Steps

Steps to reproduce this issue, eg:

1. Create a new cluster in a different DC
2. Set `primary_datacenter` to the existing DC name
3. Bootstrap ACLs on new cluster
4. Set agent replication token to token created on primary cluster
    ```
    acl = "write"

    operator = "read"

    service_prefix "" {
      policy = "read"
      intentions = "read"
    }
   ```
5. Join clusters
6. No replication

### Consul info for both Client and Server

<details>
  <summary>Existing cluster</summary>

```
consul info
agent:
        check_monitors = 0
        check_ttls = 1
        checks = 4
        services = 4
build:
        prerelease =
        revision = 40cec984
        version = 1.5.1
consul:
        acl = enabled
        bootstrap = false
        known_datacenters = 2
        leader = true
        leader_addr = 172.16.3.192:8300
        server = true
raft:
        applied_index = 10990465
        commit_index = 10990465
        fsm_pending = 0
        last_contact = 0
        last_log_index = 10990465
        last_log_term = 45
        last_snapshot_index = 10986127
        last_snapshot_term = 34
        latest_configuration = [{Suffrage:Voter ID:f33c2412-60db-15ef-ec7d-e2425c6956eb Address:172.16.3.192:8300} {Suffrage:Voter ID:b66c5bcc-06ba-7370-1ada-a833f4e58d3fAddress:172.16.4.166:8300} {Suffrage:Voter ID:d1247880-193e-ce6f-3301-4f2b06f602d1 Address:172.16.5.57:8300}]
        latest_configuration_index = 10989942
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Leader
        term = 45
runtime:
        arch = amd64
        cpu_count = 1
        goroutines = 183
        max_procs = 1
        os = linux
        version = go1.12.1
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 20
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 113
        members = 5
        query_queue = 0
        query_time = 1
serf_wan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 129
        members = 6
        query_queue = 0
        query_time = 1
```

</details>

<details>
  <summary>New cluster</summary>

```
$ consul info
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 0
        services = 0
build:
        prerelease =
        revision = 40cec984
        version = 1.5.1
consul:
        acl = enabled
        bootstrap = false
        known_datacenters = 2
        leader = true
        leader_addr = 172.16.41.27:8300
        server = true
raft:
        applied_index = 127
        commit_index = 127
        fsm_pending = 0
        last_contact = 0
        last_log_index = 127
        last_log_term = 10
        last_snapshot_index = 0
        last_snapshot_term = 0
        latest_configuration = [{Suffrage:Voter ID:3d1ac1b3-9a77-591a-64c5-b1c9e3f98953 Address:172.16.38.94:8300} {Suffrage:Voter ID:9afb079c-e298-c650-391f-41ac928e5c68Address:172.16.41.27:8300} {Suffrage:Voter ID:d7527ee6-484e-b032-2130-ae8a9792c169 Address:172.16.43.240:8300}]
        latest_configuration_index = 88
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Leader
        term = 10
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 133
        max_procs = 2
        os = linux
        version = go1.12.1
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 7
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 26
        members = 3
        query_queue = 0
        query_time = 1
serf_wan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 129
        members = 6
        query_queue = 0
        query_time = 1
```

</details>

### Operating system and Environment details

All are Amazon Linux 2 AMD64

### Log Fragments

Logs from the new cluster show what looks like a replication error. 

<details>
<summary>New cluster</summary>

```
$ journalctl -xe -u consul | grep repli
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:40:33 [INFO] raft: Added peer 3d1ac1b3-9a77-591a-64c5-b1c9e3f98953, starting replication
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:40:33 [INFO] raft: pipelining replication to peer {Voter 3d1ac1b3-9a77-591a-64c5-b1c9e3f98953 172.16.38.94:8300}
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: raft: Added peer 3d1ac1b3-9a77-591a-64c5-b1c9e3f98953, starting replication
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: raft: pipelining replication to peer {Voter 3d1ac1b3-9a77-591a-64c5-b1c9e3f98953 172.16.38.94:8300}
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:40:33 [INFO] acl: started ACL policy replication
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: acl: started ACL policy replication
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:40:33 [INFO] acl: started ACL role replication
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:40:33 [INFO] acl: started ACL token replication
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: acl: started ACL role replication
Sep 01 12:40:33 ip-172-16-41-27.ec2.internal consul[12509]: acl: started ACL token replication
Sep 01 12:40:35 ip-172-16-41-27.ec2.internal consul[12509]: replication: started Config Entry replication
Sep 01 12:40:35 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:40:35 [INFO] replication: started Config Entry replication
Sep 01 12:40:35 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:40:35 [INFO] raft: Added peer d7527ee6-484e-b032-2130-ae8a9792c169, starting replication
Sep 01 12:40:35 ip-172-16-41-27.ec2.internal consul[12509]: raft: Added peer d7527ee6-484e-b032-2130-ae8a9792c169, starting replication
Sep 01 12:40:35 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:40:35 [INFO] raft: pipelining replication to peer {Nonvoter d7527ee6-484e-b032-2130-ae8a9792c169172.16.43.240:8300}
Sep 01 12:40:35 ip-172-16-41-27.ec2.internal consul[12509]: raft: pipelining replication to peer {Nonvoter d7527ee6-484e-b032-2130-ae8a9792c169 172.16.43.240:8300}
Sep 01 12:55:30 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:55:30 [WARN] replication: Config Entry replication error (will retry if still leader): failed toretrieve remote config entries: rpc error making call: EOF
Sep 01 12:55:30 ip-172-16-41-27.ec2.internal consul[12509]: replication: Config Entry replication error (will retry if still leader): failed to retrieve remote config entries: rpc error making call: EOF
Sep 01 12:56:20 ip-172-16-41-27.ec2.internal consul[12509]: 2019/09/01 12:56:20 [WARN] replication: Config Entry replication error (will retry if still leader): failed toretrieve remote config entries: rpc error making call: EOF
Sep 01 12:56:20 ip-172-16-41-27.ec2.internal consul[12509]: replication: Config Entry replication error (will retry if still leader): failed to retrieve remote config entries: rpc error making call: EOF
```

</summary>

**Comment by ccakes** [comment]
Config files on the two clusters are pretty much identical except for `primary_datacenter`

<details>
<summary>Existing cluster - config.hcl</summary>

```
server = true
encrypt = "<snip>"
bootstrap_expect = 3
ui = true

leave_on_terminate = true
skip_leave_on_interrupt = false

enable_syslog = true
log_level = "INFO"

node_name = "hostname"
datacenter = "aws-dln"
data_dir = "/var/lib/consul/data"
retry_join = ["provider=aws tag_key=role tag_value=consul-region"]
primary_datacenter = "aws-dln"

client_addr = "0.0.0.0"

acl {
  enabled = true
  default_policy = "deny"
  down_policy = "extend-cache"
  enable_token_persistence = true
  enable_token_replication = true
}

telemetry {
  disable_hostname = true
  prometheus_retention_time = "24h"
}

watches {
  type = "service"
  service = "consul"
  passingonly = true
  handler_type = "http"

  http_handler_config {
    path = "<snip>"
    method = "POST"
    timeout = "10s"
  }
}
```

</details>

<details>
<summary>New cluster - config.hcl</summary>

```
server = true
encrypt = "<snip>"
bootstrap_expect = 3
ui = true

leave_on_terminate = true
skip_leave_on_interrupt = false

enable_syslog = true
log_level = "INFO"

node_name = "hostname"
datacenter = "aws-ash"
data_dir = "/var/lib/consul/data"
retry_join = ["provider=aws tag_key=role tag_value=consul-region"]
retry_join_wan = ["remote-consul-addr"]
primary_datacenter = "aws-dln"

client_addr = "0.0.0.0"

acl {
  enabled = true
  default_policy = "deny"
  down_policy = "extend-cache"
  enable_token_persistence = true
  enable_token_replication = true
}

telemetry {
  disable_hostname = true
  prometheus_retention_time = "24h"
}

watches {
  type = "service"
  service = "consul"
  passingonly = true
  handler_type = "http"

  http_handler_config {
    path = "<snip>"
    method = "POST"
    timeout = "10s"
  }
}
```

</details>

**Comment by ccakes** [comment]
I think I've worked out what this was - user error (poor ordering of events).

On the new cluster, `consul acl policy list` was empty even though the token list was full. I set the agent + replication token on the new cluster to a token with admin on the existing cluster and everything started syncing up properly.

Once it was done, I set the agent + replication tokens back to the previous values and so far all seems to be working as expected. I'm guessing the step I missed was creating an agent policy + token on the new cluster. Behind the scenes it was probably syncing with the old cluster but not able to write to it's own store.

Closing..
